### PR TITLE
Iss 295 validator crash when def structure not present

### PR DESCRIPTION
**Pull request by RamilCDISC**
The PR adds fix for Issue 295. A new flag has been added to select either to validate schema or not. The flag can be used with --validate_xml or -vx in test command of core.py. Default value of flag is True

To test case when schema is invalid or specifically def: structure is missing and schema validation flag is set to true. Perform following steps:

1.  Copy content of `SEND_define_DM_test_defStructure_absent.xml` and paste in the define.xml file
2. run command: `python3 -m core test -s sendig -v 3.1 -r tests/resources/CoreIssue295/SEND4.json -dp tests/resources/CoreIssue295/dm.json -dv 2.1`
3. A error message due to failed schema validation should be printed during the execution
4. verify that the summary page in the generated file contains schema validation error

To test case when schema is invalid or specifically def: structure is missing and schema validation flag is set to False. Perform following steps:

1.  Copy content of `SEND_define_DM_test_defStructure_absent.xml` and paste in the define.xml file
2. run command: `python3 -m core test -s sendig -v 3.1 -r tests/resources/CoreIssue295/SEND4.json -dp tests/resources/CoreIssue295/dm.json -dv 2.1 -vx no`
3. No error message should be printed even though schema validation fails
4. verify that the summary page in the generated file is empty and the rule in the rules report page is marked skipped

For crosscheck the case when schema is valid perform following steps:

1.  Copy content of `SEND_define_DM_test_defStructure_present.xml` and paste in the define.xml file
2. run command: `python3 -m core test -s sendig -v 3.1 -r tests/resources/CoreIssue295/SEND4.json -dp tests/resources/CoreIssue295/dm.json -dv 2.1`
3. verify that the summary page in the generated file do not contains any schema validation error

### CODE REVIEW

File: cdisc_rules_engine/dataset_builders/base_dataset_builder.py [hunk 1]
```diff
@@ -54,19 +55,22 @@ def build(self) -> pd.DataFrame:
...
-            dataset: pd.DataFrame = self.data_service.concat_split_datasets(
-                func_to_call=self.build,
-                dataset_names=self.get_corresponding_datasets_names(),
-                **kwargs,
-            )
-        else:
-            # single dataset. the most common case
-            dataset: pd.DataFrame = self.build()
-        return dataset
+        try:
```
Comment (nhaydel) [inline cdisc_rules_engine/dataset_builders/base_dataset_builder.py:58]: This block should not perform any schema validation. Schema validation should be handled in the define dataset builders

File: scripts/test_rule.py [hunk 2]
```diff
@@ -69,6 +69,7 @@ def validate_single_rule(
         whodrug_path=args.whodrug,
         define_xml_path=args.define_xml_path,
         library_metadata=library_metadata,
+        validate_xml=args.validate_xml,
```
Comment (nhaydel) [inline scripts/test_rule.py:72]: Does the run_validation script already handle this argument?

**Review by nhaydel — CHANGES_REQUESTED** [review]
Left some comments. Also needs unit tests for handling exceptions and base dataset builder changes

File: scripts/test_rule.py (hunk 2 above)
Comment (RamilCDISC) [reply scripts/test_rule.py:72]: yes, manually ran checked the validate function and also all tests related to validate passed

File: cdisc_rules_engine/dataset_builders/base_dataset_builder.py (hunk 1 above)
Comment (nhaydel) [reply cdisc_rules_engine/dataset_builders/base_dataset_builder.py:58]: This should not be changed. We should especially not be converting all exceptions to FailedSchemaValidation execeptions

File: cdisc_rules_engine/services/define_xml/base_define_xml_reader.py [hunk 3]
```diff
@@ -75,11 +76,14 @@ def read(self) -> List[dict]:
...
-        for domain_metadata in metadata.ItemGroupDef:
-            output.append(self._get_metadata_representation(domain_metadata))
-        return output
+        try:
+            metadata = self._odm_loader.MetaDataVersion()
+            output = []
+            for domain_metadata in metadata.ItemGroupDef:
+                output.append(self._get_metadata_representation(domain_metadata))
+            return output
+        except Exception:
```
Comment (nhaydel) [inline cdisc_rules_engine/services/define_xml/base_define_xml_reader.py:85]: We should include the exception message in the FailedSchemaValidation message

File: cdisc_rules_engine/services/define_xml/base_define_xml_reader.py [hunk 4]
```diff
@@ -89,15 +93,20 @@ def extract_domain_metadata(self, domain_name: str = None) -> dict:
...
+            domain_metadata_dict: dict = self._get_metadata_representation(
+                domain_metadata
+            )
+            domain_metadata_dict["define_dataset_variables"] = [
+                item_mapping.get(item.ItemOID).Name
+                for item in domain_metadata.ItemRef
+                if item.ItemOID in item_mapping
+            ]
+        except ValueError:
+            raise FailedSchemaValidation("Schema Validation Failed")
```
Comment (nhaydel) [inline cdisc_rules_engine/services/define_xml/base_define_xml_reader.py:109]: Same as above.

File: cdisc_rules_engine/services/define_xml/base_define_xml_reader.py [hunk 5]
```diff
@@ -106,16 +115,23 @@ def extract_variables_metadata(self, domain_name: str = None) -> List[dict]:
...
+                    item for item in metadata.ItemDef if item.OID == itemref.ItemOID
+                ]
+                if itemdef:
+                    variables_metadata.append(
+                        self._get_item_def_representation(
+                            itemdef[0], itemref, codelist_map
+                        )
+                    )
+        except ValueError:
+            raise FailedSchemaValidation("Schema Validation Failed")
```
Comment (nhaydel) [inline cdisc_rules_engine/services/define_xml/base_define_xml_reader.py:134]: Same as above.

File: cdisc_rules_engine/services/define_xml/base_define_xml_reader.py [hunk 6]
```diff
@@ -337,36 +358,38 @@ def get_domain_key_sequence(self, domain_name: str) -> List[str]:
...
+                    key_variables = self._get_key_variables_for_domain(
+                        domain_metadata, item_mapping
+                    )
+                    [
+                        heappush(heap, key_variable)
+                        for key_variable in key_variables
+                        if key_variable[1] not in existing_key_variables
+                    ]
+        except ValueError:
+            raise FailedSchemaValidation("Schema Validation Failed")
```
Comment (nhaydel) [inline cdisc_rules_engine/services/define_xml/base_define_xml_reader.py:392]: Same as above

File: cdisc_rules_engine/rules_engine.py [hunk 7]
```diff
@@ -174,9 +176,13 @@ def validate_single_rule(
                 return [error_obj.to_representation()]
         except Exception as e:
             logger.trace(e, __name__)
-            logger.error(
-                f"Error occurred during validation. Error: {e}. Error message: {str(e)}"
-            )
+            if not (
```
Comment (nhaydel) [inline cdisc_rules_engine/rules_engine.py:179]: I don't think we need to conditionally log here

**Review by nhaydel — CHANGES_REQUESTED** [review]

File: cdisc_rules_engine/rules_engine.py (hunk 7 above)
Comment (RamilCDISC) [reply cdisc_rules_engine/rules_engine.py:179]: The reason to do conditional log is to make sure that when validate-xml flag is set to false, then the error message is not logged,

Comment (nhaydel) [reply cdisc_rules_engine/rules_engine.py:179]: I think it's ok if it is logged though

Comment (RamilCDISC) [reply cdisc_rules_engine/rules_engine.py]: removed condition for logging in latest commit

**Review by nhaydel — APPROVED** [review]

**Comment by gerrycampion** [comment]
QA: While doing the 3rd step, I am getting an exception on Windows. This issue may be unrelated to this pull request, so we may want to log a new issue:

```
[ERROR 2023-10-10 14:38:34,738 - console_logger.py:47] - Error occurred during validation.
                Error: 'NoneType' object has no attribute 'variables'. Error message: 'NoneType' object has no attribute 'variables'
Traceback (most recent call last):
  File "C:\Users\GerryCampion\Code\cdisc-rules-engine\cdisc_rules_engine\rules_engine.py", line 158, in validate_single_rule      
    result: List[Union[dict, str]] = self.validate_rule(
  File "C:\Users\GerryCampion\Code\cdisc-rules-engine\cdisc_rules_engine\rules_engine.py", line 224, in validate_rule
    dataset = builder.get_dataset()
  File "C:\Users\GerryCampion\Code\cdisc-rules-engine\cdisc_rules_engine\dataset_builders\base_dataset_builder.py", line 68, in get_dataset
    dataset: pd.DataFrame = self.build()
  File "C:\Users\GerryCampion\Code\cdisc-rules-engine\cdisc_rules_engine\dataset_builders\variables_metadata_with_define_dataset_builder.py", line 35, in build
    content_metadata: pd.DataFrame = self.data_service.get_variables_metadata(
  File "C:\Users\GerryCampion\Code\cdisc-rules-engine\cdisc_rules_engine\services\data_services\dummy_data_service.py", line 94, in get_variables_metadata
    for i, variable in enumerate(dataset.variables):
AttributeError: 'NoneType' object has no attribute 'variables'
  [████████████████████████████████████]  100%
Output: CORE-Report-2023-10-10T14-38-34
```

I believe the issue is caused by this line:

https://github.com/cdisc-org/cdisc-rules-engine/blob/aa03ff0a35cfa6b7f35a0cfb77d2fa603016383b/cdisc_rules_engine/services/data_services/dummy_data_service.py#L51

Note that a similar filesystem issue also exists on this line:
https://github.com/cdisc-org/cdisc-rules-engine/blob/aa03ff0a35cfa6b7f35a0cfb77d2fa603016383b/cdisc_rules_engine/services/data_services/dummy_data_service.py#L46

**Comment by gerrycampion** [comment]
> QA: While doing the 3rd step, I am getting an exception on Windows. This issue may be unrelated to this pull request, so we may want to log a new issue:
> 
> ```
> [ERROR 2023-10-10 14:38:34,738 - console_logger.py:47] - Error occurred during validation.
>                 Error: 'NoneType' object has no attribute 'variables'. Error message: 'NoneType' object has no attribute 'variables'
> Traceback (most recent call last):
>   File "C:\Users\GerryCampion\Code\cdisc-rules-engine\cdisc_rules_engine\rules_engine.py", line 158, in validate_single_rule      
>     result: List[Union[dict, str]] = self.validate_rule(
>   File "C:\Users\GerryCampion\Code\cdisc-rules-engine\cdisc_rules_engine\rules_engine.py", line 224, in validate_rule
>     dataset = builder.get_dataset()
>   File "C:\Users\GerryCampion\Code\cdisc-rules-engine\cdisc_rules_engine\dataset_builders\base_dataset_builder.py", line 68, in get_dataset
>     dataset: pd.DataFrame = self.build()
>   File "C:\Users\GerryCampion\Code\cdisc-rules-engine\cdisc_rules_engine\dataset_builders\variables_metadata_with_define_dataset_builder.py", line 35, in build
>     content_metadata: pd.DataFrame = self.data_service.get_variables_metadata(
>   File "C:\Users\GerryCampion\Code\cdisc-rules-engine\cdisc_rules_engine\services\data_services\dummy_data_service.py", line 94, in get_variables_metadata
>     for i, variable in enumerate(dataset.variables):
> AttributeError: 'NoneType' object has no attribute 'variables'
>   [████████████████████████████████████]  100%
> Output: CORE-Report-2023-10-10T14-38-34
> ```
> 
> I believe the issue is caused by this line:
> 
> https://github.com/cdisc-org/cdisc-rules-engine/blob/aa03ff0a35cfa6b7f35a0cfb77d2fa603016383b/cdisc_rules_engine/services/data_services/dummy_data_service.py#L51
> 
> Note that a similar filesystem issue also exists on this line:
> 
> https://github.com/cdisc-org/cdisc-rules-engine/blob/aa03ff0a35cfa6b7f35a0cfb77d2fa603016383b/cdisc_rules_engine/services/data_services/dummy_data_service.py#L46

Update, I see line 51 is going to be fixed here: https://github.com/cdisc-org/cdisc-rules-engine/pull/540

And line 46 is not used

**Pull request merged**